### DISCUSSION TITLE
What's the reasoning behind preventing the use of UTMCTL over SSH?

### DISCUSSION THREAD

**Discussion by pkolyvas** [utmapp/UTM#6397] [General]
## Use Case
I use UTM as direct replacement for Vagrant on my Apple Silicon macs. 

I run headless arm64 builds of Linux servers which I can SSH into for local development/testing purposes. 

Typically, I'll setup a default "base" image and then clone it—not unlike using templates/machine images in cloud platforms like AWS.

Recently I ran into an issue where a number of docker images for specific workflow weren't available for arm64 platforms. Instead of attempting to rebuild those images as multi-arch or specifically for arm64, I thought I'd log into my intel mac workstation via SSH and use `utmctl` in the same way there to expedite development. 

## Problem

When I tried to run `utmctl` over SSH I got the following message:
```
Error from event: The operation couldn’t be completed. (OSStatus error -1743.)
NOTE: utmctl does not work from SSH sessions or before logging in.
```
I wanted to prompt a discussion on why this is the case? Particularly for headless VMs, it strikes me as an uncessaary limitation, but again, I'm asking from a place where I probably don't have the whole picture. 

The strange thing is, there's an easy workaround: simply starting a "screen" session directly from the remote mac, and then accessing it remotely allows me to use `utmctl` over SSH just fine so the limitation seems moot? Or maybe my use case is limited and the workaround is fine.

**Comment by naveenrajm7** [comment]
Hello 

Thanks for sharing the workaround, it's really helpful as I am trying to see if I can support [vagrant share](https://developer.hashicorp.com/vagrant/docs/share) for the  [Vagrant UTM Plugin](https://github.com/naveenrajm7/vagrant_utm) ( Check out the [plugin](https://naveenrajm7.github.io/vagrant_utm/) if you want vagrant + UTM setup)

Also, I found the reason why `utmctl` doesn't work over ssh from https://github.com/utmapp/UTM/issues/5377#issuecomment-1653147303
>This is unfortunately a limitation of how apple scripting bridge works. You must have the GUI launched or launchable in an interactive instance. I don’t think it works with SSH.

**Comment by pkolyvas** [comment]
Will totally check out the Vagrant UTM plugin! Didn't know it exists. Thank you. :)

**Comment by mcx808** [comment]
> The strange thing is, there's an easy workaround: simply starting a "screen" session directly from the remote mac, and then accessing it remotely allows me to use `utmctl` over SSH just fine so the limitation seems moot? Or maybe my use case is limited and the workaround is fine.

Do you mean you just launched screen (`/usr/bin/screen`) from within the ssh session then called `utmctl`? I tried that but I still get the same error when trying to list or start VMs.

**Reply by pkolyvas** [reply]
Ahh, great question that made me realise I wasn't clear. 

I have to launch screen _locally_ on the target machine. So for example, sitting in front of my Mac I start a named screen session.

Then I can remotely ssh into that mac, connect to the screen session, and use `utmctl` via the CLI.

Put another way, I can't start a screen session _remotely_, but I can connect to an existing screen session remotely provided that the session was started locally.

Reading it back it's all still kind of confusing. I hope I've been somewhat clear. 

Sorry about that.

**Reply by mcx808** [reply]
No problem, and thanks for the quick response! Since I'm not using vagrant and want to start a UTM VM over ssh I'll leave a screen session running and use your workaround.

**Reply by naveenrajm7** [reply]
Hey.. 
just seeking feedback.
Since both of you seem to use UTM remotely, would something like UTM Web be of any interest to you?
What would be the top 3 features you'd like to see in UTM Web for your use case. 

UTM web demo : https://github.com/utmapp/UTM/issues/6767#issuecomment-2562320395

**Reply by mcx808** [reply]
In this case I only need to start a VM, so for UTM web if we can also make an API call to do the same action then that would work too.

**Reply by javimontero** [reply]
`tmux` also works as a workaround.

**Reply by mcx808** [reply]
> Hey.. just seeking feedback. Since both of you seem to use UTM remotely, would something like UTM Web be of any interest to you? What would be the top 3 features you'd like to see in UTM Web for your use case.
> 
> UTM web demo : [#6767 (comment)](https://github.com/utmapp/UTM/issues/6767#issuecomment-2562320395)

Is there any progress on this? The screen workaround isn't working for me anymore on macOS 26.
